sansa-rdf-spark/src/main/scala/net/sansa_stack/rdf/spark/model/df/TripleOps.scala

```diff
@@ -0,0 +1,309 @@
+import org.apache.spark.sql.types.{ StructField, StructType, StringType }
+
+/**
+ * Spark/DataFrame based implementation of DataFrame of triples.
+ *
+ * @author Gezim Sejdiu
+ */
+
+object TripleOps {
+
+  /**
+   * Convert a [[DataFrame]] into a RDD[Triple].
+   *
+   * @param triples DataFrame of triples (as string).
+   * @return a RDD of triples.
+   */
+  def toRDD(triples: DataFrame): RDD[Triple] = {
+    triples.rdd.map(row =>
+      Triple.create(
+        NodeFactory.createURI(row.getString(0)),
+        NodeFactory.createURI(row.getString(1)),
+        //NodeFactory.createLiteral(row.getString(2))
+        if (row.getString(2).startsWith("http:")) NodeFactory.createURI(row.getString(2)) else NodeFactory.createLiteral(row.getString(2))))
+  }
+
+  /**
+   * Convert a DataFrame of Triple into a Dataset of Triple.
+   *
+   * @param triples DataFrame of triples.
+   * @return a Dataset of triples.
+   */
+  def toDS(triples: DataFrame): Dataset[Triple] = {
+    val spark: SparkSession = SparkSession.builder().getOrCreate()
+    implicit val encoder = Encoders.kryo[Triple]
+    //triples.as[Triple]
+    spark.createDataset[Triple](toRDD(triples))
+  }
+
+  /**
+   * Get triples.
+   *
+   * @param triples DataFrame of triples.
+   * @return DataFrame which contains list of the triples.
+   */
+  def getTriples(triples: DataFrame): DataFrame =
+    triples
+
+  /**
+   * Get subjects.
+   *
+   * @param triples DataFrame of triples.
+   * @return DataFrame which contains list of the subjects.
+   */
+  def getSubjects(triples: DataFrame): DataFrame =
+    triples.select("s")
+
+  /**
+   * Get predicates.
+   *
+   * @param triples DataFrame of triples.
+   * @return DataFrame which contains list of the predicates.
+   */
+  def getPredicates(triples: DataFrame): DataFrame =
+    triples.select("p")
+
+  /**
+   * Get objects.
+   *
+   * @param triples DataFrame of triples.
+   * @return DataFrame which contains list of the objects.
+   */
+  def getObjects(triples: DataFrame): DataFrame =
+    triples.select("o")
+
+  /**
+   * Returns an DataFrame of triples that match with the given input.
+   *
+   * @param triples DataFrame of triples
+   * @param subject the subject
+   * @param predicate the predicate
+   * @param object the object
+   * @return DataFrame of triples
+   */
+  def find(triples: DataFrame, subject: Option[String] = None, predicate: Option[String] = None, `object`: Option[String] = None): DataFrame = {
+
+    var sql = s"SELECT s, p, o FROM TRIPLES"
+
+    if (subject.isDefined || predicate.isDefined || `object`.isDefined) {
+      sql += " WHERE "
+
+      val conditions = scala.collection.mutable.ListBuffer[String]()
+
+      if (subject.isDefined) conditions += s"s = '${subject.get}'"
+      if (predicate.isDefined) conditions += s"p = '${predicate.get}'"
+      if (`object`.isDefined) conditions += s"o = '${`object`.get}'"
+
+      sql += conditions.mkString(" AND ")
```

Comment:
Factor out sql string generation into a separate method/function

Reply:
Thanks for the hint. I just separated that part into a separate method [here](https://github.com/SANSA-Stack/SANSA-RDF/blob/feature/df-tripleops/sansa-rdf-spark/src/main/scala/net/sansa_stack/rdf/spark/model/df/TripleOps.scala#L108).